File: src/endpoints/json/put_item.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.libs.check import check_url
from src.libs.object import json_parse, json_stringify
from src.modules.verify import checking_token

async def put_item(params: dict = {}, req = None, _db: DB = None, _token = None):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.PutItem(**params)

        # checking token
        token = checking_token(req, db) if not _token else _token

        # check parse json
        json_data = json_parse(params.json_data)

        # check category_srl
        if params.category_srl:
            from ..category import __libs__ as category_libs
            count = db.get_count(
                table_name = Table.CATEGORY.value,
                where = [
                    f'AND module LIKE \'{category_libs.Module.JSON}\'',
                    f'AND srl = {params.category_srl}',
                ],
            )
            if not (count > 0): raise Exception('Invalid category_srl', 400)

        # set values
        values = {
            'category_srl': params.category_srl or None,
            'name': params.name,
            'description': params.description or None,
            'json': json_stringify(json_data, None) or '{}',
        }

        # set placeholders
        placeholders = [
            { 'key': 'category_srl', 'value': ':category_srl' },
            { 'key': 'name', 'value': ':name' },
            { 'key': 'description', 'value': ':description' },
            { 'key': 'json', 'value': ':json' },
            { 'key': 'created_at', 'value': 'DATETIME("now", "localtime")' },
            { 'key': 'updated_at', 'value': 'DATETIME("now", "localtime")' },
        ]

        # add item
        data = db.add_item(
            table_name = Table.JSON.value,
            placeholders = placeholders,
            values = values,
        )

        # add tag
        if params.tag:
            from ..tag import __libs__ as tag_libs
            tag_libs.add(
                _db = db,
                tags = params.tag,
                module = tag_libs.Module.JSON,
                module_srl = data,
            )

        # set result
        result = output.success({
            'message': 'Complete add JSON.',
            'data': data,
        }, _req=req)
    except Exception as e:
        result = output.exc(e, _req=req)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/json/put-item.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import { parseJSON } from '@/libs/objects'
import * as tagHelper from '@/routes/tag/__helper'
import * as categoryHelper from '@/routes/category/__helper'
import type { JsonModel } from './__model'

type PutItemParams = {
  body: JsonModel['putItemBody'],
}

export default async function putItem({ body }: PutItemParams)
{
  let _transaction = false
  try
  {
    // check parse json
    const jsonData = parseJSON(body.json)

    // check category
    if (body.category)
    {
      // TODO: 작동하는지 확인 필요함
      const _count = categoryHelper.count({
        where: [
          `AND srl = ${body.category}`,
          `AND module LIKE \'${categoryHelper.MODULE.JSON}\'`,
        ],
      })
      if (_count <= 0) throw new ServiceError(`Invalid category`, { status: 400 })
    }

    // begin transaction
    _transaction = db.transaction('begin')

    // add json data
    const added = db.addData({
      table: DB.TABLE.JSON,
      values: [
        { key: 'category_srl', value: body.category },
        { key: 'name', value: body.name },
        { key: 'description', value: body.description },
        { key: 'json', value: JSON.stringify(jsonData) || '{}' },
        { key: 'created_at', valueName: DB.DATE_TIME },
        { key: 'updated_at', valueName: DB.DATE_TIME },
      ],
    })

    // add tag
    if (body.tag)
    {
      const _tags = body.tag.split(',')
      for (const tag of _tags)
      {
        tagHelper.add({
          module: tagHelper.MODULE.JSON,
          module_srl: added.data,
          tag,
        })
      }
    }

    // commit transaction
    _transaction = db.transaction('commit')

    return added.data
  }
  catch (_e: any)
  {
    // collback transaction
    db.transaction('rollback', _transaction)

    throw new ServiceError('Failed to add JSON.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
